narwhals/_pandas_like/namespace.py

```diff
@@ -250,3 +250,116 @@ def concat(
                 backend_version=self._backend_version,
             )
         raise NotImplementedError
+
+    def when(
+        self,
+        *predicates: IntoPandasLikeExpr,
+    ) -> PandasWhen:
+        plx = self.__class__(self._implementation, self._backend_version)
+        if predicates:
+            condition = plx.all_horizontal(*predicates)
+        else:
+            msg = "at least one predicate needs to be provided"
+            raise TypeError(msg)
+
+        return PandasWhen(condition, self._implementation, self._backend_version)
+
+
+def _when_then_value_arg_process(
+    plx: PandasLikeNamespace,
+    value: PandasLikeExpr | PandasLikeSeries | Any,
+    *,
+    shape: tuple[int],
+) -> PandasLikeExpr:
+    from narwhals.dependencies import get_numpy
+
+    # NumPy is a required dependency of pandas
+    np = get_numpy()
+
+    if isinstance(value, PandasLikeExpr):
+        return value
+    elif isinstance(value, PandasLikeSeries):
+        return plx._create_expr_from_series(value)
+    elif (np := get_numpy()) is not None and isinstance(value, np.ndarray):
+        return plx._create_expr_from_series(plx._create_compliant_series(value))
+    else:
+        return plx._create_expr_from_series(
+            plx._create_compliant_series(np.full(shape[0], value))
+        )
+
+
+class PandasWhen:
+    def __init__(
+        self,
+        condition: PandasLikeExpr,
+        implementation: Implementation,
+        backend_version: tuple[int, ...],
+        then_value: Any = None,
+        otherise_value: Any = None,
+    ) -> None:
+        self._implementation = implementation
+        self._backend_version = backend_version
+        self._condition = condition
+        self._then_value = then_value
+        self._otherwise_value = otherise_value
+
+    def __call__(self, df: PandasLikeDataFrame) -> list[PandasLikeSeries]:
+        from narwhals._pandas_like.namespace import PandasLikeNamespace
+
+        plx = PandasLikeNamespace(
+            implementation=self._implementation, backend_version=self._backend_version
+        )
+
+        condition = self._condition._call(df)[0]
+
+        value_series = _when_then_value_arg_process(
+            plx, self._then_value, shape=condition.shape
+        )._call(df)[0]
+        otherwise_series = _when_then_value_arg_process(
+            plx, self._otherwise_value, shape=condition.shape
+        )._call(df)[0]
+
+        return [value_series.zip_with(condition, otherwise_series)]
```

Comment:
this would create an `object` dtype series, which we need to avoid

the dtype of `value_series` needs to be preserved

Reply:
I assume we are talking about the `None` possibility for `otherwise_value`?

Reply:
yup!

Reply:
ok, looking deeper into this

we have found an issue with `zip_with` in the `pandas` implementation as in the `polars` documentation of `zip_with`, [here](https://docs.pola.rs/api/python/stable/reference/series/api/polars.Series.zip_with.html), `other` series already has to be of the same type of the series it is zipped with. No such guarantees are made in the `pandas` implementation in narwhals, which leads to this issue.

Reply:
I have added a fix for this in this pr for now, if you want we can move this fix into a separate pr

Reply:
Also note that polars also does some sneaky casting sometimes, example

```python
import polars as pl

s = pl.Series([1, 2, 3])
df = pl.DataFrame({'s': s})

print(df)
# shape: (3, 1)
# ┌─────┐
# │ s   │
# │ --- │
# │ i64 │
# ╞═════╡
# │ 1   │
# │ 2   │
# │ 3   │
# └─────┘





print(df.select(pl.when(pl.col('s') == 1).then(-1).alias('s_cast')))
# shape: (3, 1)
# ┌────────┐
# │ s_cast │
# │ ---    │
# │ i32    │
# ╞════════╡
# │ -1     │
# │ null   │
# │ null   │
# └────────┘
```

Reply:
> ok, looking deeper into this

> we have found an issue with zip_with in the pandas implementation as in the polars documentation of zip_with, [here](https://docs.pola.rs/api/python/stable/reference/series/api/polars.Series.zip_with.html), other series already has to be of the same type of the series it is zipped with. No such guarantees are made in the pandas implementation in narwhals, which leads to this issue.

I have removed this fix now, though the issue still stands, I have replaced it with an expansion on the `maybe_cast_dtypes` functionality

Reply:
Now the functionality is as follows

```python
import narwhals as nw
import pandas as pd
import polars as pl


@nw.narwhalify
def function(df):
    return df.with_columns(nw.when(nw.col('s') == 1).then(-1).alias('s_when'))

s = [1, 2, 3, 4, 5]
df_pd = pd.DataFrame({'s': s})
df_pl = pl.DataFrame({'s': s})





print(function(df_pd))
# s  s_when
# 0  1      -1
# 1  2    <NA>
# 2  3    <NA>
# 3  4    <NA>
# 4  5    <NA>





print(function(df_pd).dtypes)
# s         int64
# s_when    Int64





print(function(df_pl))
# dtype: object
# shape: (5, 2)
# ┌─────┬────────┐
# │ s   ┆ s_when │
# │ --- ┆ ---    │
# │ i64 ┆ i32    │
# ╞═════╪════════╡
# │ 1   ┆ -1     │
# │ 2   ┆ null   │
# │ 3   ┆ null   │
# │ 4   ┆ null   │
# │ 5   ┆ null   │
# └─────┴────────┘
```

Reply:
thanks for looking into this so deep!

i'd like to spend a little more time on this - nullable dtypes aren't generally supported by the data ecosystem unfortunately, so i'm not sure we should convert to them on behalf of the user

we'll get this in 💪 let's just take our time to avoid having to make breaking changes later

Reply:
I had other thoughts on how to solve this for the type issue, which again mean changing the type of the column in pandas

- changing to a `float` type
- use masked arrays for  the`integer` case